### ISSUE TITLE
Support for custom namespaces

### ISSUE THREAD

**Issue by iuyiuy** [SVG-Edit/svgedit#94]
I have a need for having some elements with some added meaning, that I would like to manage myself. This would normally be done with namespaces, but there doesn't seem to be a way to add namespaces and elements. It's possible to add to the svgedit.NS object from the config file, but this won't allow the namespace declaration in the svg tag, nor the actual elements using this namespace.

I could always manage the custom namespaces, either in extensions or from the output afterwards.

**Comment by praneybehl** [comment]
I support that :+1:

**Comment by flintobrien** [comment]
We use an extension to add a custom namespace (see example below). There is one bug that prevents this from working. See [pull request #110](https://github.com/SVG-Edit/svgedit/pull/110).

#### Example Extension:
```
svgedit.editor.addExtension( "ext-fisheye-base", function ( S ) {
   svgedit.sanitize.addWhiteListException("g", 'http://www.totalgrid.org', 'some-attribute');
   svgedit.sanitize.addWhiteListException("g", 'http://www.totalgrid.org', 'another-attribute');
   svgedit.sanitize.addWhiteListException("rect", 'http://www.totalgrid.org', 'some-attribute');
   svgedit.sanitize.addWhiteListException("rect", 'http://www.totalgrid.org', 'another-attribute');
   ...
}
```

**Comment by flintobrien** [comment]
I just realized that [sanitize.addWhiteListException()](https://github.com/gec/svgedit/blob/fisheye/editor/sanitize.js#L105) is only in the [GEC fork, branch fisheye](https://github.com/gec/svgedit/tree/fisheye) of svgedit. I can do a pull request if others think it would be helpful as part of a more complete solution.

**Comment by iuyiuy** [comment]
I implemented this in pull request no. 97:
https://github.com/SVG-Edit/svgedit/pull/97

**Comment by flintobrien** [comment]
The ```svgedit.ignoredNS``` solution in [#97](https://github.com/SVG-Edit/svgedit/pull/97) is interesting. What are people's thoughts for the alternative solution of extending the whitelist either as a callable API or config?

Here's an example config:
```
svgEditor.setConfig({
  extendedNamespaces: {
    MYNS: 'http://mynamespace.com'
  },
  extendedWhiteList: {
    "g": ["MYNS:some-attribute", "MYNS:another-attribute"],
    "rect": ["MYNS:some-attribute", "MYNS:another-attribute"]
  }
});
```

**Comment by iuyiuy** [comment]
The ignoredNS solution gives you the freedom to not have to extend the config with every element in the added namespace. It could be a long list and it might even be changing often or maybe it's not very well documented. If svg-edit isn't going to use it, i don't see a reason for having to specify it.

**Comment by distler** [comment]
> The ignoredNS solution gives you the freedom to not have to extend the config with every element in the added namespace.

Having to declare an explicit whitelist of the allowed elements and attributes is a *feature*, not a bug.

It would be bad to call `ignoredNS` on a namespace which contained (say) `<foo:script>`, even if your extension didn't use it.

You also needs to be cautious about `extendedWhiteList` in `setConfig`. We *don't* want the end-user to be able to call

```
http://your.site.com/svg-edit/editor/svg-editor.html?extendedWhiteList["script"]='""
```

**Comment by flintobrien** [comment]
Seems like there are multiple related issues to resolve. Can people comment so we can work toward a consensus.

1. Extending the whitelist would be a good feature. I don't see any objections so far.
1. Should svgedit support a list of "unmanaged" namespaces (aka. ingnoreNS).
1. Should the "se" namespace be unmanaged. I know "se" attributes aren't in the whitelist now, but I believe they should be (according to the inferred design and "TODO" comments). The only use I've found in the current code is **se:nonce** attribute. I'd like to hear from someone who knows the history of NS.SE and it's intended use.
1. Should the "se" namespace be used by extensions or should each extension have it's own namespace? Currently, extensions could be using it and no one would know.

**Comment by flintobrien** [comment]
Seeking clarification on _distler's_ comment...

> You also need to be cautious about extendedWhiteList in setConfig. We don't want the end-user to be able to call
```
http://your.site.com/svg-edit/editor/svg-editor.html?extendedWhiteList["script"]="type"
```

Are you saying:
* Extending the whitelist should not be available as a query string option.
* Certain elements should not allow extended attributes (ex: "script").
* other?

**Comment by distler** [comment]
>Are you saying:
>
>* Extending the whitelist should not be available as a query string option.
>* Certain elements should not allow extended attributes (ex: "script").
>* other?

The first.

* #2 is a blacklist, rather than a whitelist.
* End-users should not be able to modify the whitelist (set by site policy).

> The only use I've found in the current code is se:nonce attribute. I'd like to hear from someone who knows the history of NS.SE and it's intended use.

My understanding is that NS.SE was intended for internal use by SVG-Edit, and is mostly stripped-out on output.

I was the one who lobbied for `se.nonce`, both the attribute itself *and* for it not to be stripped-out on output. The purpose is that if an HTML page contains *several* SVG figures, drawn in SVG-Edit:

1. The `id`s should be unique across the page.
2. It should be possible to re-edit a figure, preserving the previously-assigned `id`s.

The nonce ensures (with reasonably high probability) uniqueness. Preserving the attribute on output ensures the ability to re-edit.

**Comment by distler** [comment]
As far as I can tell, the only other attribute currently in NS.SE is `se:connector` (though there may be extensions that also use that namespace).

**Comment by iuyiuy** [comment]
I can see the appeal of whitelisting tags in each namespace, but as the [W3C proposal for inline SVG in HTML](https://dev.w3.org/SVG/proposals/svg-html/svg-html-proposal.html) - and possibly others - say in the requirements section:
> Should allow for unrestricted growth of the SVG language by the SVG specifications (though those specifications should also take into account the idea that SVG will, going forward, be used more commonly in concert with HTML). This means that there would be no "white list" of allowed SVG elements in HTML.

There will be changes to various standards and even if we don't know how to handle them, I don't think we should remove them. Especially if someone wants to add extra namespaces.

The mention of the script tag, makes it seem that the concern is security and we should address whether there could be one. My suggestion was just to add a custom namespace and allowing foo:script shouldn't allow html:script or svn:script. If the browser doesn't recognize the namespace, I believe it will ignore it. But please correct me if I'm wrong. This means it won't be a problem in this specific scenario.

I don't see a reason to allow this to be set in the URL, but I currently only see a couple of options for limiting URL given settings and they pretty much boil down to limiting either SVG content, extensions or everything. Do you have a particular way you'd want this set? Should we add a preventURLCustomNamespaces and have it default to true or should we not allow it at all?

Outside the scope of this issue, [SVG itself allows script tags](https://www.w3.org/TR/SVG/script.html). That should definitely not be allowed with SVG content given through the query string, but it is part of the SVG standard.

**Comment by distler** [comment]
1. I think we need to distinguish between what is allowed in the *language* and what should be allowed in a web application which (at least in some use-cases) allows untrusted content to be displayed to end-users. The `<script>` element is *without question* a part of the HTML language. That doesn't mean that Github should allow it in this comment-thread.

2. If you allow custom namespaces, then are you are going to allow custom namespace declarations, too? If so, then what about
   ```
   xmlns:foo="http://www.w3.org/1999/xhtml" foo:onclick="..."
   ```
   (among many other variations on the theme).

3. Again, I have no objection to a mechanism by which the *site* (that includes extension authors) -- but *not* the end-user -- can extend the whitelist . But I still think that, on security grounds, it should remain a whitelist (where the elements and attributes to be added are explicitly named).

**Comment by iuyiuy** [comment]
As I said in my comment, this should obviously be restricted from the URL settings. I wasn't sure if there would be a need to have the option to enable it or just parts of it through the config.js file. I'll update my patch to limit it completely and then it can always be expanded to be an option if a need arises.

But even the click event is [supported in SVG officially](https://www.w3.org/TR/SVG/interact.html). I'm not saying these scripts should be executed by default, but I could see editing them being a feature in the future. Right now I just needed the ability to whitelist my own namespace, though. Which is what I implemented in pull request #97. It doesn't have a whitelist for specific attributes, but that could be added as a further option.

**Comment by iuyiuy** [comment]
I've just pushed a commit to pull request #97, which removes the ability to set ignoredNS through the query string.

**Comment by codedread** [comment]
I looked over the pull request and didn't see any issue (though I didn't
understand why you decide to lower-case all the NS keys).  Jacques, would
you mind taking a look at https://github.com/SVG-Edit/svgedit/pull/97/files
 ?

Jeff

**Comment by iuyiuy** [comment]
I'll reply to the case issue in the pull request, in order to not duplicate the discussion.

**Comment by codedread** [comment]
+Jacques Distler <distler@golem.ph.utexas.edu> any chance you can look at
KB's latest patch at
https://github.com/SVG-Edit/svgedit/pull/97#issuecomment-247574819 ?

**Comment by distler** [comment]
1. Definitely **do not** change the namespace URI for the `NS.SE` namespace. Namespace URIs don't change every time you relocated your code repository. That way lies madness.
2. The `NS.SE` namespace should, by default, be in `svgedit.ignoredNS` and the `se` prefix in `svgedit.ignoredNSUsedAlias`(to retain the previous behaviour).
3. I would like to see some tests included, to show that this code does what it's advertised to do.

**Comment by iuyiuy** [comment]
1. Alright. I figured since the URL didn't include versioning and assuming there have been changes to the namespace, that the fact that the repo moved is enough for a new 'version'. A bit like how the Linux Kernel turned to version 3.0 at some point, mostly because it was about time for a new major version number. But I'll change it back.

2. I thought the `se` namespace was supposed to be explicitly filtered by allowed tags and attributes. But I guess we could ignore it.

3. Are you aware that the existing tests didn't pass, at least in Firefox? I opened issue #128 for that. Of course any new tests won't necessarily have to be dependent on the existing ones, but it does indicate there might be some generic issues that need to be sorted out.
